File: core/hamming/impl/Hamming.cs
/// LadybugDB Hamming Operations - C#
/// Same XOR + POPCOUNT as Python, TypeScript, Rust, Go, C...

using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Security.Cryptography;
using System.Text;

namespace LadybugDB
{
    public class HammingVector
    {
        public const int DIM = 10_000;
        public const int DIM_U64 = 157;
        public const ulong LAST_MASK = (1UL << 16) - 1;

        public ulong[] Data { get; }

        public HammingVector() { Data = new ulong[DIM_U64]; }

        public HammingVector(ulong[] data) { Data = data; }

        public static HammingVector FromSeed(string seed)
        {
            var data = new ulong[DIM_U64];
            using var sha256 = SHA256.Create();
            
            for (int i = 0; i < DIM_U64; i++)
            {
                var input = Encoding.UTF8.GetBytes($"{seed}:{i}");
                var hash = sha256.ComputeHash(input);
                data[i] = BitConverter.ToUInt64(hash, 0);
            }
            data[DIM_U64 - 1] &= LAST_MASK;
            return new HammingVector(data);
        }

        public HammingVector Xor(HammingVector other)
        {
            var result = new ulong[DIM_U64];
            for (int i = 0; i < DIM_U64; i++)
                result[i] = Data[i] ^ other.Data[i];
            result[DIM_U64 - 1] &= LAST_MASK;
            return new HammingVector(result);
        }

        public int Hamming(HammingVector other)
        {
            int total = 0;
            for (int i = 0; i < DIM_U64; i++)
                total += BitOperations.PopCount(Data[i] ^ other.Data[i]);
            return total;
        }

        public double Similarity(HammingVector other) => 1.0 - (double)Hamming(other) / DIM;

        public static HammingVector operator ^(HammingVector a, HammingVector b) => a.Xor(b);
    }

    public static class Ladybug
    {
        public static HammingVector Fingerprint(string name, string signature, string body)
            => HammingVector.FromSeed($"{name}::{signature}::{body}");

        public static List<(int Index, double Similarity)> Resonate(
            HammingVector query, IList<HammingVector> corpus, double threshold = 0.5)
        {
            return corpus
                .Select((v, i) => (Index: i, Similarity: query.Similarity(v)))
                .Where(x => x.Similarity >= threshold)
                .OrderByDescending(x => x.Similarity)
                .ToList();
        }
    }
}

File: core/hamming/impl/hamming.go
// 10K Hamming Operations - Go
package hamming

import (
    "math/bits"
    "sort"
)

const (
    DIM     = 10000
    DIM_U64 = 157
    LAST_MASK = (1 << 16) - 1
)

type Vector [DIM_U64]uint64

func Hamming(a, b *Vector) int {
    total := 0
    for i := 0; i < DIM_U64; i++ {
        total += bits.OnesCount64(a[i] ^ b[i])
    }
    return total
}

func Similarity(a, b *Vector) float64 {
    return 1.0 - float64(Hamming(a, b))/float64(DIM)
}

func XorBind(a, b *Vector) Vector {
    var result Vector
    for i := 0; i < DIM_U64; i++ {
        result[i] = a[i] ^ b[i]
    }
    result[DIM_U64-1] &= LAST_MASK
    return result
}

func BatchHamming(query *Vector, corpus []Vector) []int {
    results := make([]int, len(corpus))
    for i := range corpus {
        results[i] = Hamming(query, &corpus[i])
    }
    return results
}

type Match struct {
    Index      int
    Similarity float64
}

func Resonate(query *Vector, corpus []Vector, threshold float64) []Match {
    var results []Match
    for i := range corpus {
        sim := Similarity(query, &corpus[i])
        if sim >= threshold {
            results = append(results, Match{i, sim})
        }
    }
    sort.Slice(results, func(i, j int) bool {
        return results[i].Similarity > results[j].Similarity
    })
    return results
}